### DISCUSSION TITLE
a simple html plugin for the last page of the task

### DISCUSSION THREAD

**Discussion by xf15** [jspsych/jsPsych#1138] [Q&A] [answered]
You guys are providing many powerful and useful plugins but i don't find a simple html only one that doesn't take any response. It's just for the last page of the task where I show people their code, the url for the next task, and where i say thank you. My current solution is that I use a html-button plugin and make the key choice weird like delete so they don't accidentally press it and end up on a blank page. It looks weird since the progress bar is not completely filled. Is there a better solution?

**Comment by becky-gilbert** [comment, marked as answer]
I suggest just using the [html-keyboard-response plugin](https://www.jspsych.org/plugins/jspsych-html-keyboard-response/), since that allows you to put HTML-formatted text on the screen without any on-screen response options, and you can ignore any key presses by using `jsPsych.NO_KEYS` as the value for the `choices` parameter.

And for the progress bar, you can manually set that to 100% using the [`jsPsych.setProgressBar` function](https://www.jspsych.org/core_library/jspsych-core/#jspsychsetprogressbar). I think you could do this in the trial's [on_load function](https://www.jspsych.org/overview/callbacks/#on_load). Here's an example:
```javascript
var end_page = {
  type: 'html-keyboard-response',
  choices: jsPsych.NO_KEYS,
  stimulus: '<p>All done!</p>',
  on_load: function() {
    jsPsych.setProgressBar(1);
  }
};
```

**Reply by xf15** [reply]
right right, i was just reading on manually setting the progress bar: because it reflects the top level timeline only i need to manually set it anyway. thank you @becky-gilbert

**Reply by xf15** [reply]
I just want to point out that as I am hosting the task on cognition.run, with the last trial being html-keyboard-response plugin and its choices being jsPsych.NO_KEYS or some key that subjects don't press, this trial will not be recorded in the csv that i can download from cognition.run, and the run will be marked unfinished.

**Reply by becky-gilbert** [reply]
I see what you mean about wanting a final 'thank you' page that doesn't require the participant to do anything, but where the data is already saved at that point. I normally do this by putting my data-saving function inside the on_load function for this last 'thank you' page. But with cognition.run, I don't think there's an option to 'manually' save the data at any point other than the end of the experiment. You could get in touch with the cognition.run team to find out if it's possible for you to call the data-saving function in your script. 

You could also ask cognition.run whether it's possible to pass a callback function to their data-saving function. In the callback function, you could redirect the participant to a different cognition.run 'study', which would really be just a page that just says "Thanks" etc. That way the data would be saved before the participant sees the final thank you page.

Apart from that, the only other suggestions I have are to (1) use a button response plugin for the last page, and label the button with something that the participant is likely to click, like 'Finish experiment' or 'Save responses', or (2) include a trial duration for the last page, so that the data saves automatically when the trial times out (but you wouldn't want to use a long duration for this, because otherwise you might get participants who close the browser window before the experiment finishes).

**Reply by xf15** [reply]
sorry that I was not being clear that there was no problem for me anymore, just wanted to share it as a note. With cognition.run, the data is saved continuously as the subject is making progress with the task. I implemented what you suggested, all data are saved except the last trial which has nothing I want to save. So it's all good.
